File: handlers/submitbasket.go
package handlers

import (
	"encoding/json"
	"github.com/adrianfish/dialang-web/data"
	"github.com/adrianfish/dialang-web/datacapture"
	"github.com/adrianfish/dialang-web/models"
	"github.com/adrianfish/dialang-web/scoring"
	"github.com/adrianfish/dialang-web/session"
	"log"
	"net/http"
	"net/url"
	"slices"
	"strconv"
	"strings"
)

func SubmitBasket(w http.ResponseWriter, r *http.Request) {

	dialangSession := session.SessionManager.Get(r.Context(), "session").(models.DialangSession)

	if dialangSession.TES.TL == "" || dialangSession.TES.Skill == "" || dialangSession.CurrentBasketId == 0 {
		log.Println("None of the test language, skill or current basket id were set in the session. Returning 500 ...")
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	basketType := r.FormValue("basketType")
	if basketType == "" {
		log.Println("No basketType supplied. Returning 400 (Bad Request) ...")
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	positionInBasketSorter := func(a, b *models.ScoredItem) int {
		return a.PositionInBasket - b.PositionInBasket
	}

	numScoredItems := len(dialangSession.ScoredItems)
	currentBasketId := dialangSession.CurrentBasketId

	log.Printf("basketType: %v\n", basketType)
	log.Printf("currentBasketId: %v\n", currentBasketId)
	log.Printf("currentBasketNumber: %v\n", dialangSession.CurrentBasketNumber)
	log.Printf("scored item list length: %v\n", numScoredItems)

	returnMap := map[string]any{}

	itemList := dialangSession.ScoredItems

	switch basketType {
	case "mcq":
		{
			log.Println("mcq")
			itemId, itemIdErr := strconv.Atoi(r.FormValue("itemId"))
			if itemIdErr != nil {
				log.Println("Invalid or missing item id. Returning 400 (Bad Request) ...")
				w.WriteHeader(http.StatusBadRequest)
				return
			}
			answerId, responseIdErr := strconv.Atoi(r.FormValue("response"))
			if responseIdErr != nil {
				log.Println("No response supplied. Returning 400 (Bad Request) ...")
				w.WriteHeader(http.StatusBadRequest)
				return
			}

			if scoredItem, err := scoring.GetScoredIdResponseItem(itemId, answerId); err != nil {
				log.Printf("Failed to get scored item for itemId %d and responseId %d\n", strconv.Itoa(itemId), strconv.Itoa(answerId))
				w.WriteHeader(http.StatusInternalServerError)
				return
			} else {
				scoredItem.BasketId = dialangSession.CurrentBasketId
				scoredItem.PositionInBasket = 1
				scoredItem.ResponseId = answerId
				scoredItem.PositionInTest = numScoredItems + 1
				log.Printf("Item position in test: %d\n", scoredItem.PositionInTest)
				scoredItem.Answers = data.ItemAnswers[itemId]
				itemList = append(itemList, scoredItem)
				scoredBasket := models.ScoredBasket{Id: currentBasketId, Type: "mcq", Skill: scoredItem.Item.Skill, Items: []*models.ScoredItem{scoredItem}}
				dialangSession.ScoredBaskets = append(dialangSession.ScoredBaskets, &scoredBasket)
				returnMap["scoredBasket"] = scoredBasket
				datacapture.LogSingleIdResponse(dialangSession.PassId, scoredItem)
			}
		}

	case "tabbedpane":
		{
			itemsToLog := []*models.ScoredItem{}
			responses := getMultipleIdResponses(&r.Form)
			basketItems := []*models.ScoredItem{}
			for itemId, answerId := range responses {
				if item, err := scoring.GetScoredIdResponseItem(itemId, answerId); err == nil {
					item.ResponseId = answerId
					position := r.FormValue(strconv.Itoa(item.Item.Id) + "-position")
					if position != "" {
						if positionInBasket, err := strconv.Atoi(position); err == nil {
							item.PositionInBasket = positionInBasket
						} else {
							log.Printf("Failed to convert positionInBasket to int: %w\n", err)
						}
					} else {
						log.Printf("No position supplied for item '%d. Returning 400 (Bad Request) ...\n", item.Item.Id)
						w.WriteHeader(http.StatusBadRequest)
						return
					}

					item.BasketId = currentBasketId
					item.PositionInTest = numScoredItems + item.PositionInBasket
					log.Printf("Item position in basket: %d\n", item.PositionInBasket)
					log.Printf("Item position in test: %d\n", item.PositionInTest)
					log.Println(item.BasketId)
					item.Answers = data.ItemAnswers[item.Item.Id]
					itemsToLog = append(itemsToLog, item)
					itemList = append(itemList, item)
					basketItems = append(basketItems, item)
				} else {
					log.Println("No item returned from scoring")
				}
			}
			slices.SortFunc(basketItems, positionInBasketSorter)
			scoredBasket := models.ScoredBasket{Id: currentBasketId, Type: "tabbedpane", Skill: basketItems[0].Item.Skill, Items: basketItems}
			dialangSession.ScoredBaskets = append(dialangSession.ScoredBaskets, &scoredBasket)
			returnMap["scoredBasket"] = scoredBasket
			datacapture.LogMultipleIdResponses(dialangSession.PassId, itemsToLog)
		}

	case "shortanswer":
		{
			responses := getMultipleTextualResponses(&r.Form)
			basketItems := []*models.ScoredItem{}
			itemsToLog := []*models.ScoredItem{}
			for itemId, responseText := range responses {
				if item, err := scoring.GetScoredTextResponseItem(itemId, responseText); err == nil {
					item.BasketId = currentBasketId
					item.ResponseText = responseText
					position := r.FormValue(strconv.Itoa(item.Item.Id) + "-position")
					if position != "" {
						if positionInBasket, err := strconv.Atoi(position); err == nil {
							item.PositionInBasket = positionInBasket
						} else {
							log.Printf("Failed to convert positionInBasket to int: %s\n", err)
						}
					} else {
						log.Printf("No position supplied for item '%d. Returning 400 (Bad Request) ...\n", item.Item.Id)
						w.WriteHeader(http.StatusBadRequest)
						return
					}
					item.PositionInTest = numScoredItems + item.PositionInBasket
					log.Printf("Item position in test: %d\n", item.PositionInTest)
					item.Answers = data.ItemAnswers[item.Item.Id]
					itemList = append(itemList, item)
					itemsToLog = append(itemsToLog, item)
					basketItems = append(basketItems, item)
				} else {
					log.Println("No item returned from scoring")
				}
			}
			slices.SortFunc(basketItems, positionInBasketSorter)
			scoredBasket := models.ScoredBasket{Id: currentBasketId, Type: "shortanswer", Skill: basketItems[0].Item.Skill, Items: basketItems}
			dialangSession.ScoredBaskets = append(dialangSession.ScoredBaskets, &scoredBasket)
			returnMap["scoredBasket"] = scoredBasket
			datacapture.LogMultipleTextualResponses(dialangSession.PassId, itemsToLog)
		}

	case "gaptext":
		{
			responses := getMultipleTextualResponses(&r.Form)
			basketItems := []*models.ScoredItem{}
			itemsToLog := []*models.ScoredItem{}

			for itemId, responseText := range responses {
				if item, err := scoring.GetScoredTextResponseItem(itemId, responseText); err == nil {
					item.BasketId = currentBasketId
					item.ResponseText = responseText
					position := r.FormValue(strconv.Itoa(item.Item.Id) + "-position")
					if position != "" {
						if positionInBasket, err := strconv.Atoi(position); err == nil {
							item.PositionInBasket = positionInBasket
						} else {
							log.Printf("Failed to convert positionInBasket to int: %s\n", err)
						}
					} else {
						log.Printf("No position supplied for item '%d. Returning 400 (Bad Request) ...\n", item.Item.Id)
						w.WriteHeader(http.StatusBadRequest)
						return
					}
					item.PositionInTest = numScoredItems + item.PositionInBasket
					log.Printf("Item position in test: %d\n", item.PositionInTest)
					item.Answers = data.ItemAnswers[item.Item.Id]
					itemList = append(itemList, item)
					itemsToLog = append(itemsToLog, item)
					basketItems = append(basketItems, item)
				} else {
					log.Println("No item returned from scoring")
				}
			}
			slices.SortFunc(basketItems, positionInBasketSorter)
			scoredBasket := models.ScoredBasket{Id: currentBasketId, Type: "gaptext", Skill: basketItems[0].Item.Skill, Items: basketItems}
			dialangSession.ScoredBaskets = append(dialangSession.ScoredBaskets, &scoredBasket)
			returnMap["scoredBasket"] = scoredBasket
			datacapture.LogMultipleTextualResponses(dialangSession.PassId, itemsToLog)
		}

	case "gapdrop":
		{
			responses := getMultipleIdResponses(&r.Form)
			basketItems := []*models.ScoredItem{}
			itemsToLog := []*models.ScoredItem{}
			for itemId, responseId := range responses {
				if item, err := scoring.GetScoredIdResponseItem(itemId, responseId); err == nil {
					item.BasketId = currentBasketId
					item.ResponseId = responseId

					position := r.FormValue(strconv.Itoa(item.Item.Id) + "-position")
					if position != "" {
						if positionInBasket, err := strconv.Atoi(position); err == nil {
							item.PositionInBasket = positionInBasket
						} else {
							log.Printf("Failed to convert positionInBasket to int: %s\n", err)
							w.WriteHeader(http.StatusBadRequest)
							return
						}
					} else {
						log.Printf("No position supplied for item '%d. Returning 400 (Bad Request) ...\n", item.Item.Id)
						w.WriteHeader(http.StatusBadRequest)
						return
					}

					item.PositionInTest = numScoredItems + item.PositionInBasket
					log.Printf("Item position in test: %d\n", item.PositionInTest)
					item.Answers = data.ItemAnswers[item.Item.Id]
					itemList = append(itemList, item)
					itemsToLog = append(itemsToLog, item)
					basketItems = append(basketItems, item)
				} else {
					log.Printf("No item returned from scoring: %w\n", err)
				}
			}

			slices.SortFunc(basketItems, positionInBasketSorter)
			scoredBasket := models.ScoredBasket{Id: currentBasketId, Type: "gapdrop", Skill: basketItems[0].Item.Skill, Items: basketItems}
			dialangSession.ScoredBaskets = append(dialangSession.ScoredBaskets, &scoredBasket)
			returnMap["scoredBasket"] = scoredBasket
			datacapture.LogMultipleIdResponses(dialangSession.PassId, itemsToLog)
		}

	default:
		{
			log.Println("Unrecognised basketType supplied. Returning 400 (Bad Request) ...")
			w.WriteHeader(http.StatusBadRequest)
			return
		}
	}

	dialangSession.ScoredItems = itemList

	nextBasketNumber := dialangSession.CurrentBasketNumber + 1
	log.Printf("nextBasketNumber: %d\n", nextBasketNumber)

	basketIds := data.BookletBaskets[dialangSession.BookletId]

	if nextBasketNumber >= len(basketIds) {
		// The test has finished. Grade it.
		rawScore, itemGrade, itemLevel := scoring.GetItemGrade(dialangSession.TES.TL,
			dialangSession.TES.Skill,
			dialangSession.BookletId,
			itemList)

		log.Printf("ITEM GRADE: %d\n", itemGrade)

		dialangSession.ItemRawScore = rawScore
		dialangSession.ItemGrade = itemGrade
		dialangSession.ItemLevel = itemLevel

		session.SessionManager.Put(r.Context(), "session", dialangSession)

		datacapture.LogTestResult(&dialangSession)
		datacapture.LogTestFinish(dialangSession.PassId)

		if dialangSession.ResultUrl != "" {
			parts := strings.Split(dialangSession.ResultUrl, "?")
			var params strings.Builder
			if len(parts) == 2 {
				params.WriteString("?")
				params.WriteString(parts[1])
				params.WriteString("&")
			} else {
				params.WriteString("?")
			}
			params.WriteString("itemGrade=")
			params.WriteString(strconv.Itoa(itemGrade))
			if dialangSession.SaLevel != "" {
				params.WriteString("&saLevel=")
				params.WriteString(dialangSession.SaLevel)
			}
			if dialangSession.VsptLevel != "" {
				params.WriteString("&vsptLevel=")
				params.WriteString(dialangSession.VsptLevel)
			}
			url := parts[0] + params.String()
			log.Printf("Redirect URL: %v\n", url)
			returnMap["redirect"] = url
			json.NewEncoder(w).Encode(returnMap)
		} else {
			// We set testDone to true so the client js knows to enable the sa feedback and advice buttons
			returnMap["itemLevel"] = itemLevel
			returnMap["testDone"] = "true"
			json.NewEncoder(w).Encode(returnMap)
		}
	} else {
		datacapture.LogBasket(dialangSession.PassId, currentBasketId, dialangSession.CurrentBasketNumber)

		nextBasketId := basketIds[nextBasketNumber]

		dialangSession.CurrentBasketNumber = nextBasketNumber
		dialangSession.CurrentBasketId = nextBasketId
		session.SessionManager.Put(r.Context(), "session", dialangSession)

		returnMap["nextBasketId"] = nextBasketId
		returnMap["itemsCompleted"] = len(itemList)

		json.NewEncoder(w).Encode(returnMap)
	}
}

func getMultipleIdResponses(params *url.Values) map[int]int {

	responses := map[int]int{}

	for k, v := range *params {
		if !strings.HasSuffix(k, "-response") {
			continue
		}
		if itemId, err := strconv.Atoi(strings.Split(k, "-")[0]); err == nil {
			log.Println("Failed to convert item id to int", err)
			if answerId, err := strconv.Atoi(v[0]); err == nil {
				responses[itemId] = answerId
			} else {
				log.Println("Failed to convert answer id to int", err)
			}
		} else {
			log.Println("Failed to convert item id to int", err)
		}
	}
	return responses
}

/**
 * Returns a map of response text on to itemId
 */
func getMultipleTextualResponses(params *url.Values) map[int]string {

	responses := map[int]string{}

	for k, v := range *params {
		if !strings.HasSuffix(k, "-response") {
			continue
		}

		if itemId, err := strconv.Atoi(strings.Split(k, "-")[0]); err == nil {
			responses[itemId] = v[0]
		} else {
			log.Println("Failed to convert item id to int", err)
		}
	}

	return responses
}

File: src/routes/submitbasket.ts
import { Context } from "@hono/hono";
import { getSessionId } from "../utils/utils.ts";
import { Storage } from "../storage/storage.ts";
import { LTIKit, Score } from "@adrianfish/lti-kit";
import type { DialangSession, ScoredBasket, ScoredItem, TES } from "../types.ts";
import { getItemGrade, getScoredIdResponseItem, getScoredTextResponseItem } from "../scoring/scoring.ts";

export async function submitBasket(
  c: Context,
  storage: Storage,
  lti: LTIKit,
): Promise<Response> {

  const sessionId: string | undefined = getSessionId(c);
  if (!sessionId) {
		console.error("Failed to get session id");
		c.status(500);
		return c.html("");
  }

  const session: DialangSession | null = await storage.getSession(sessionId);

  if (!session) {
		console.error(`No session for id ${sessionId}`);
		c.status(500);
		return c.html("");
  }

	if (!session.tl || !session.skill || !session.currentBasketId || !session.bookletId || session.currentBasketNumber === undefined) {
		console.error("None of the test language, skill, current basket id, booklet id or current basket number were set in the session. Returning 500 ...");
		c.status(500);
		return c.html("");
	}

  const body = await c.req.parseBody();

  const basketType: string = body["basketType"] as string;
  if (!basketType) {
    console.warn("No basketType supplied. Returning 400 (Bad Request) ...")
    c.status(400);
    return c.html("");
  }

  const positionInBasketSorter = (a: ScoredItem, b: ScoredItem) => a.positionInBasket - b.positionInBasket;

  const numScoredItems: number = session.scoredItems?.length || 0;
  const currentBasketId: number = session.currentBasketId;

  console.debug(`basketType: ${basketType}`);
  console.debug(`currentBasketId: ${currentBasketId}`);
  console.debug(`currentBasketNumber: ${session.currentBasketNumber}`);
  console.debug(`scored item list length: ${numScoredItems}`);

  const returnMap: Record<string, any> = {};

  const itemList = session.scoredItems || [];

  switch (basketType) {
    case "mcq": {
      const itemId: number = parseInt(body["itemId"] as string);
      if (!itemId) {
        console.error("Invalid or missing item id. Returning 400 (Bad Request) ...")
        c.status(400);
        return c.html("");
      }
      const answerId: number = parseInt(body["response"] as string);
      if (!answerId) {
        console.warn("No response supplied. Returning 400 (Bad Request) ...")
        c.status(400);
        return c.html("");
      }

      const [ scoredItem, err ] = await getScoredIdResponseItem(itemId, answerId, storage);
      if (err) {
        console.error(`Failed to get scored item for itemId ${itemId} and responseId ${answerId}`);
        c.status(500);
        return c.html("");
      } else {
        scoredItem.basketId = session.currentBasketId;
        scoredItem.positionInBasket = 1;
        scoredItem.responseId = answerId;
        scoredItem.positionInTest = numScoredItems + 1;
        console.debug(`Item position in test: ${scoredItem.positionInTest}`);
        scoredItem.answers = await storage.getItemAnswers(itemId);
        itemList.push(scoredItem);
        const scoredBasket: ScoredBasket = { id: currentBasketId, type: "mcq", skill: scoredItem.skill, items: [scoredItem] };
        returnMap["scoredBasket"] = scoredBasket
        //datacapture.LogSingleIdResponse(session.passId, scoredItem)
      }
      break;
    }

    case "tabbedpane": {
      const itemsToLog: Array<ScoredItem> = [];
      const responses: Record<number, number> = getMultipleIdResponses(body);
      const basketItems: Array<ScoredItem> = [];
      const entries = Object.entries(responses);
      for (let i = 0; i < entries.length; i++) {
        const [ itemId, answerId ] = entries[i];
        const [ item, error ] = await getScoredIdResponseItem(parseInt(itemId), answerId, storage);
        item.responseId = answerId;
        const position: string = body[item.id + "-position"] as string;
        if (position) {
          item.positionInBasket = parseInt(position);
        } else {
          console.warn(`No position supplied for item ${item.id}. Returning 400 (Bad Request) ...`);
          c.status(400);
          return c.html("");
        }

        item.BasketId = currentBasketId;
        item.positionInTest = numScoredItems + item.positionInBasket;
        console.debug(`Item position in basket: ${item.positionInBasket}`);
        console.debug(`Item position in test: ${item.PositionInTest}`);
        item.answers = await storage.getItemAnswers(item.id);
        itemsToLog.push(item);
        itemList.push(item);
        basketItems.push(item);
      }
      basketItems.sort(positionInBasketSorter);
      const scoredBasket: ScoredBasket = { id: currentBasketId, type: "tabbedpane", skill: basketItems[0].skill, items: basketItems };
      returnMap["scoredBasket"] = scoredBasket;
      //datacapture.LogMultipleIdResponses(session.passId, itemsToLog)
      break;
    }

    case "shortanswer": {
      const responses = getMultipleTextualResponses(body);
      const basketItems: Array<ScoredItem> = [];
      const itemsToLog: Array<ScoredItem> = [];
      const entries = Object.entries(responses);
      for (let i = 0; i < entries.length; i++) {
        const [ itemId, responseText ] = entries[i];
        const [ item, error ] = await getScoredTextResponseItem(parseInt(itemId), responseText, storage);
        if (item) {
          item.basketId = currentBasketId;
          item.responseText = responseText;
          const position: string = body[item.id + "-position"] as string;
          if (position) {
            item.positionInBasket = parseInt(position);
          } else {
            console.warn(`No position supplied for item ${item.id}. Returning 400 (Bad Request) ...`);
            c.status(400);
            return c.html("");
          }
          item.positionInTest = numScoredItems + item.positionInBasket;
          console.debug(`Item position in test: ${item.positionInTest}`);
          item.answers = await storage.getItemAnswers(item.id);
          itemList.push(item);
          itemsToLog.push(item);
          basketItems.push(item);
        } else {
          console.error("No item returned from scoring");
        }
      }
      basketItems.sort(positionInBasketSorter);
      const scoredBasket: ScoredBasket = { id: currentBasketId, type: "shortanswer", skill: basketItems[0].skill, items: basketItems };
      returnMap["scoredBasket"] = scoredBasket;
      //datacapture.LogMultipleTextualResponses(session.passId, itemsToLog)
      break;
    }

    case "gaptext": {
      const responses = getMultipleTextualResponses(body);
      const basketItems: Array<ScoredItem> = [];
      const itemsToLog: Array<ScoredItem> = [];

      const entries = Object.entries(responses);
      for (let i = 0; i < entries.length; i++) {
        const [ itemId, responseText ] = entries[i];

        const [ item, error ] = await getScoredTextResponseItem(parseInt(itemId), responseText, storage);
        if (item) {
          item.basketId = currentBasketId;
          item.responseText = responseText;
          const position: string = body[item.id + "-position"] as string;
          if (position) {
            item.positionInBasket = parseInt(position);
          } else {
            console.warn(`No position supplied for item ${item.id}. Returning 400 (Bad Request) ...`);
            c.status(400);
            return c.html("");
          }
          item.positionInTest = numScoredItems + item.positionInBasket;
          console.debug(`Item position in test: ${item.positionInTest}`);
          item.answers = await storage.getItemAnswers(item.id);
          itemList.push(item);
          itemsToLog.push(item);
          basketItems.push(item);
        } else {
          console.error("No item returned from scoring");
        }
      }
      basketItems.sort(positionInBasketSorter);
      const scoredBasket: ScoredBasket = { id: currentBasketId, type: "gaptext", skill: basketItems[0].skill, items: basketItems };
      returnMap["scoredBasket"] = scoredBasket;
      break;
      //datacapture.LogMultipleTextualResponses(session.passId, itemsToLog)
    }

    case "gapdrop": {
      const responses: Record<number, number> = getMultipleIdResponses(body);
      const basketItems: Array<ScoredItem> = [];
      const itemsToLog: Array<ScoredItem> = [];

      const entries = Object.entries(responses);
      for (let i = 0; i < entries.length; i++) {
        const [ itemId, responseId ] = entries[i];
        const [ item, error ] = await getScoredIdResponseItem(parseInt(itemId), responseId, storage);

        if (item) {
          item.basketId = currentBasketId;
          item.responseId = responseId;

          const position: string = body[item.id + "-position"] as string;
          if (position) {
            item.positionInBasket = parseInt(position);
          } else {
            console.warn(`No position supplied for item ${item.id}. Returning 400 (Bad Request) ...`);
            c.status(400);
            return c.html("");
          }

          item.positionInTest = numScoredItems + item.positionInBasket;
          item.answers = await storage.getItemAnswers(item.id);
          itemList.push(item);
          itemsToLog.push(item);
          basketItems.push(item);
        } else {
          console.error("No item returned from scoring");
        }
      }

      basketItems.sort(positionInBasketSorter);
      const scoredBasket: ScoredBasket = { id: currentBasketId, type: "gapdrop", skill: basketItems[0].skill, items: basketItems };
      returnMap["scoredBasket"] = scoredBasket;
      break;
      //datacapture.LogMultipleIdResponses(session.passId, itemsToLog)
    }

    default: {
      console.warn("Unrecognised basketType supplied. Returning 400 (Bad Request) ...");
      c.status(400);
      return c.html("");
    }
  }

  session.scoredItems = sparsifyItems(itemList);

  const nextBasketNumber = session.currentBasketNumber + 1;
  console.debug(`nextBasketNumber: ${nextBasketNumber}`);

  const basketIds: Array<number> | null = await storage.getBaskets(session.bookletId);

  if (!basketIds) {
    console.error(`No basket ids for booklet ${session.bookletId}`);
    c.status(500);
    return c.html("");
  }

  if (nextBasketNumber >= basketIds.length) {
    // The test has finished. Grade it.
    const [ rawScore, itemGrade, itemLevel ]
      = await getItemGrade(session.tl, session.skill, session.bookletId, itemList, storage);

    // If we're in an LTI session, post the score.
    if (session.isLTI) {
      const ltik: string = body["ltik"] as string;
      const userId = session.user;

      if (ltik && userId) {
        const score: Score = {
          userId,
          scoreGiven: rawScore,
          scoreMaximum: 1000,
          activityProgress: "Completed",
          gradingProgress: "NotReady",
        };
        // Use the null lineItemUrl version so the single line item is picked up by the lti lib.
        lti.postScore(ltik, null, score);
      } else {
        console.warn("We're in an LTI session, but the ltik or userId were not supplied. Score not posted");
      }
    }

    session.itemRawScore = rawScore;
    session.itemGrade = itemGrade;
    session.itemLevel = itemLevel;

    storage.saveSession(sessionId, session);

    //datacapture.LogTestFinish(session.passId)
    storage.logTestResult(session);

    if (session.resultUrl) {
      const parts = session.resultUrl.split("?");
      let url = `${parts[0]}?`;
      if (parts.length == 2) {
        url += `${parts[1]}&`;
      }
      url += `itemGrade=${itemGrade}`;
      if (session.saLevel) {
        url += `&saLevel=${session.saLevel}`;
      }
      if (session.vsptLevel) {
        url += `&vsptLevel=${session.vsptLevel}`;
      }
      returnMap["redirect"] = url;
      return c.json(returnMap);
    } else {
      // We set testDone to true so the client js knows to enable the sa feedback and advice buttons
      returnMap["itemLevel"] = itemLevel;
      returnMap["testDone"] = "true";
      return c.json(returnMap);
    }
  } else {
    //datacapture.LogBasket(session.passId, currentBasketId, session.currentBasketNumber)

    const nextBasketId = basketIds[nextBasketNumber];

    session.currentBasketNumber = nextBasketNumber;
    session.currentBasketId = nextBasketId;
    storage.saveSession(sessionId, session);

    returnMap["nextBasketId"] = nextBasketId;
    returnMap["itemsCompleted"] = itemList.length;

    return c.json(returnMap);
  }
}

function getMultipleIdResponses(body: object): Record<number, number> {

  const responses: Record<number, number> = [];

  Object.entries(body).forEach(([k, v]) => {

    if (!k.endsWith?.("-response")) return;

    const itemId: number = parseInt(k.split("-")[0]);
    const answerId: number = parseInt(v as string);
    responses[itemId] = answerId;
  });
  return responses;
}

/**
 * Returns a map of response text on to itemId
 */
function getMultipleTextualResponses(body: object): Record<number, string> {

  const responses: Record<number, string> = {};

  Object.entries(body).forEach(([k, v]) => {

    if (!k.endsWith("-response")) return;

    const itemId: number = parseInt(k.split("-")[0]);
    responses[itemId] = v as string;
  });

  return responses;
}

function sparsifyItems(items: Array<ScoredItem>): Array<any> {
  return items.map(item => ({ id: item.id, score: item.score }));
}
